src/chrome/chromeDebugAdapter.ts

```diff
@@ -449,32 +453,39 @@ export abstract class ChromeDebugAdapter implements IDebugAdapter {
      * This event tells the client to begin sending setBP requests, etc. Some consumers need to override this
      * to send it at a later time of their choosing.
      */
-    protected sendInitializedEvent(): void {
+    protected async sendInitializedEvent(): Promise<void> {
         // Wait to finish loading sourcemaps from the initial scriptParsed events
         if (this._initialSourceMapsP) {
             const initialSourceMapsP = this._initialSourceMapsP;
             this._initialSourceMapsP = null;
 
-            initialSourceMapsP.then(() => {
+            await initialSourceMapsP.then(async () => {
                 this._session.sendEvent(new InitializedEvent());
-                this._earlyScripts.forEach(script => this.sendLoadedSourceEvent(script));
+                await Promise.all(this._earlyScripts.map(async script => await this.sendLoadedSourceEvent(script)));
                 this._earlyScripts = null;
             });
         }
     }
 
+    public doAfterProcessingSourceEvents(action: () => void): Promise<void> {
+        return this._sourceLoadedQueue = this._sourceLoadedQueue.then(action);
+    }
+
     /**
      * e.g. the target navigated
      */
-    private onExecutionContextsCleared(): void {
-        const asyncOperations = [];
-        this._scriptsById.forEach(scriptedParseEvent => {
-            asyncOperations.push(this.scriptToLoadedSourceEvent('removed', scriptedParseEvent).then(scriptEvent =>
-                this._session.sendEvent(scriptEvent)));
-        });
+    private onExecutionContextsCleared(): Promise<void> {
+            const asyncOperations = [];
+            this._scriptsById.forEach(scriptedParseEvent => {
+                return this.doAfterProcessingSourceEvents(async () => { // This will not execute until all the on-flight 'new' source events have been processed
+                    asyncOperations.push(this.scriptToLoadedSourceEvent('removed', scriptedParseEvent).then(scriptEvent => {
+                        this._session.sendEvent(scriptEvent);
+                    }));
+                });
+            });
 
-        Promise.all(asyncOperations).then(() =>
-            this.clearTargetContext());
+            return Promise.all(asyncOperations).then(() =>
```

Comment:
This is confusing, `asyncOperations` will not be populated by this point because it's populated by an async function that we don't wait on.

Reply:
Nice catch... I missed this...